### PR TITLE
add trivy scanner

### PR DESCRIPTION
**Pull request by andriisoldatenko**
**What this PR does / why we need it**:

fix https://github.com/astronomer/issues/issues/743

**Special notes for your reviewer**:
from official [trivy docs](https://github.com/aquasecurity/trivy#vs-clair):
> Trivy collects vulnerability information in Alpine Linux from Alpine Linux aports repository. Then, those vulnerabilities will be saved on vuln-list.
alpine-secdb has 6959 vulnerabilities (as of 2019/05/12). vuln-list has 11101 vulnerabilities related to Alpine Linux (as of 2019/05/12). There is a difference in detection accuracy because the number of vulnerabilities is nearly doubled.
In addition, Trivy analyzes the middle layers as well to find out which version of the library was used for static linking.
Clair can not handle the following cases because it analyzes the image after applying all layers.```

### CODE REVIEW

**Comment by andriisoldatenko** [comment]
@ashb PTAL, also we need to cherry-pick few commits from upstream?
https://github.com/apache/airflow/pull/7280

**Comment by andriisoldatenko** [comment]
or we can whitelist 1 issue and merge?

**Comment by andriisoldatenko** [comment]
@ashb i can whitelist issue for all except 1.10.9?

**Comment by andriisoldatenko** [comment]
```
#!/bin/sh -eo pipefail
trivy --ignore-unfixed -s HIGH,CRITICAL --exit-code 1 --no-progress "astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild"
2020-03-10T11:03:44.765Z	INFO	Need to update DB
2020-03-10T11:03:44.765Z	INFO	Downloading DB...
2020-03-10T11:03:45.909Z	INFO	Reopening DB...
2020-03-10T11:04:03.164Z	INFO	Detecting Alpine vulnerabilities...
2020-03-10T11:04:03.165Z	INFO	Detecting yarn vulnerabilities...
astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild (alpine 3.10.4)
==================================================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
usr/lib/python3.7/site-packages/airflow/www_rbac/yarn.lock
==========================================================
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)
+---------+------------------+----------+-------------------+---------------+--------------------------------+
| LIBRARY | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |             TITLE              |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
| lodash  | CVE-2018-16487   | HIGH     | 3.10.1            | >=4.17.11     | lodash: Prototype pollution in |
|         |                  |          |                   |               | utilities function             |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
Exited with code exit status 1
```

Upstream airflow is already patched





1:10
https://github.com/apache/airflow/pull/7280

**Comment by kaxil** [comment]
> ```
> #!/bin/sh -eo pipefail
> trivy --ignore-unfixed -s HIGH,CRITICAL --exit-code 1 --no-progress "astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild"
> 2020-03-10T11:03:44.765Z	INFO	Need to update DB
> 2020-03-10T11:03:44.765Z	INFO	Downloading DB...
> 2020-03-10T11:03:45.909Z	INFO	Reopening DB...
> 2020-03-10T11:04:03.164Z	INFO	Detecting Alpine vulnerabilities...
> 2020-03-10T11:04:03.165Z	INFO	Detecting yarn vulnerabilities...
> astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild (alpine 3.10.4)
> ==================================================================
> Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)
> usr/lib/python3.7/site-packages/airflow/www_rbac/yarn.lock
> ==========================================================
> Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)
> +---------+------------------+----------+-------------------+---------------+--------------------------------+
> | LIBRARY | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |             TITLE              |
> +---------+------------------+----------+-------------------+---------------+--------------------------------+
> | lodash  | CVE-2018-16487   | HIGH     | 3.10.1            | >=4.17.11     | lodash: Prototype pollution in |
> |         |                  |          |                   |               | utilities function             |
> +---------+------------------+----------+-------------------+---------------+--------------------------------+
> Exited with code exit status 1
> ```
> 
> Upstream airflow is already patched
> 
> 1:10
> [apache/airflow#7280](https://github.com/apache/airflow/pull/7280)

We have those cherry-picked to our **v1-10-7** version https://github.com/astronomer/airflow/commits/v1-10-7

Do you want me to tag and create a new version for it @ashb ?

**Comment by kaxil** [comment]
I have created new tag https://github.com/astronomer/airflow/commits/v1.10.7+astro.7

Build: https://travis-ci.com/github/astronomer/airflow/builds/153061412

Once that passes we can create a new PR to updated 1.10.7 folder in this repo

**Comment by ashb** [comment]
Thanks Kaxil.

**Comment by andriisoldatenko** [comment]
@ashb we are good to go!

File: .circleci/config.yml [hunk 1]
```diff
@@ -313,6 +369,51 @@ jobs:
...
+          command: |
+            apk add --update-cache --upgrade curl rpm
+            curl -sfL https://raw.githubusercontent.com/aquasecurity/trivy/master/contrib/install.sh | sh -s -- -b /usr/local/bin
+      - restore_cache:
+          keys:
+            - trivy-cache
+      - run:
+          name: Scan the local image with trivy
+          command: |
+            cat \<<EOF > .trivyignore
```
Comment (ashb) [inline .circleci/config.yml]: What did we do for Clair ignore file? My preference would be to have this directly as a file in the repo

**Comment by andriisoldatenko** [comment]
@ashb PTAL

File: .circleci/config.yml (hunk 1 above)
Comment (andriisoldatenko) [reply .circleci/config.yml]: fixed.

File: .trivyignore [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# ignore in old airflow versions; it's fixed ONLY in new airflow version
+# lodash: Prototype pollution in utilities function fixed in >=4.17.11
+CVE-2018-16487
```
Comment (ashb) [inline .trivyignore]: Hmmmm, not sure about a global ignore here -- if it somehow comes back in 1.10.7 we'd want to know about it.

That probably means we need more than one of these files (a shared one, and one per airflow version (or per image?)). Hmmm

File: .circleci/config.yml [hunk 3]
```diff
@@ -313,6 +376,44 @@ jobs:
...
+          command: |
+            apk add --update-cache --upgrade curl rpm
+            curl -sfL https://raw.githubusercontent.com/aquasecurity/trivy/master/contrib/install.sh | sh -s -- -b /usr/local/bin
+      - restore_cache:
+          keys:
+            - trivy-cache
+      - run:
+          name: Scan the local image with trivy
+          command: |
+            trivy --ignorefile ./<< parameters.airflow_version >>/<< parameters.distribution >>/.trivyignore --cache-dir /tmp/workspace/trivy-cache --ignore-unfixed -s HIGH,CRITICAL --exit-code 1 --no-progress "astronomerinc/ap-airflow:<< parameters.airflow_version >>-<< parameters.distribution_name >>"
```
Comment (ashb) [inline .circleci/config.yml]: Since we are specifying it, let's make the file `trivyignore` instead of `.trivyignore`

File: .circleci/config.yml [hunk 4]
```diff
@@ -313,6 +376,44 @@ jobs:
...
+      - save_cache:
+          key: trivy-cache
+          paths:
+            - /tmp/workspace/
```
Comment (ashb) [inline .circleci/config.yml]: This is going to include the docker images that were sand too - we should only save `/tmp/workspace/trivy-cache`

File: 1.10.7/alpine3.10/.trivyignore [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# ignore in old airflow versions; it's fixed ONLY in new airflow version
+# lodash: Prototype pollution in utilities function fixed in >=4.17.11
+CVE-2018-16487
```
Comment (ashb) [inline 1.10.7/alpine3.10/.trivyignore]: Do we still need this one for 1.10.7? I forget whether we pulled the UI fixes back to.

Comment (andriisoldatenko) [reply 1.10.7/alpine3.10/.trivyignore]:
```
#!/bin/sh -eo pipefail
trivy --ignorefile ./1.10.7/alpine3.10/trivyignore --cache-dir /tmp/workspace/trivy-cache --ignore-unfixed -s HIGH,CRITICAL --exit-code 1 --no-progress "astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild"
2020-03-16T20:08:58.762Z	INFO	Need to update DB
2020-03-16T20:08:58.762Z	INFO	Downloading DB...
2020-03-16T20:08:59.663Z	INFO	Reopening DB...
2020-03-16T20:09:22.366Z	INFO	Detecting Alpine vulnerabilities...
2020-03-16T20:09:22.367Z	INFO	Detecting yarn vulnerabilities...

astronomerinc/ap-airflow:1.10.7-alpine3.10-onbuild (alpine 3.10.4)
==================================================================
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)


usr/lib/python3.7/site-packages/airflow/www_rbac/yarn.lock
==========================================================
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 1, CRITICAL: 0)

+---------+------------------+----------+-------------------+---------------+--------------------------------+
| LIBRARY | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |             TITLE              |
+---------+------------------+----------+-------------------+---------------+--------------------------------+
| lodash  | CVE-2018-16487   | HIGH     | 3.10.1            | >=4.17.11     | lodash: Prototype pollution in |
|         |                  |          |                   |               | utilities function             |
+---------+------------------+----------+-------------------+---------------+--------------------------------+

Exited with code exit status 1
```

File: .circleci/config.yml (hunk 4 above)
Comment (andriisoldatenko) [reply .circleci/config.yml]: fixed.

File: .circleci/config.yml (hunk 3 above)
Comment (andriisoldatenko) [reply .circleci/config.yml]: done.

**Comment by andriisoldatenko** [comment]
@ashb PTAL

**Pull request merged**